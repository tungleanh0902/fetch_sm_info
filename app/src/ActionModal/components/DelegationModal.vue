<template>
  <ActionModal
    id="delegation-modal"
    ref="actionModal"
    :validate="validateForm"
    :amount="isRedelegation ? 0 : amount"
    :title="isRedelegation ? 'Restake' : 'Stake'"
    class="delegation-modal"
    submission-error-prefix="Staking failed"
    :transaction-type="isRedelegation ? messageType.RESTAKE : messageType.STAKE"
    :transaction-data="transactionData"
    :notify-message="notifyMessage"
    feature-flag="delegate"
    :disabled="
      isInElection ||
      (hasNoNominations && Object.keys(targetValidator).length === 0)
    "
    @close="clear"
    @txIncluded="onSuccess"
  >
    <TmFormGroup
      v-if="session.addressRole === `stash`"
      class="action-modal-form-group"
    >
      <div class="form-message notice">
        <span>
          This is a stash account, you can increase the amount to stake but you
          need to sign in with your controller account to set or change your
          validators.
        </span>
      </div>
    </TmFormGroup>
    <TmFormGroup
      v-if="session.addressRole === `controller`"
      class="action-modal-form-group"
    >
      <div class="form-message notice">
        <span>
          This is a controller account, you can set or change your validators
          but to increase the amount to stake you need to sign in with your
          stash account.
        </span>
      </div>
    </TmFormGroup>
    <TmFormGroup class="action-modal-form-group">
      <div class="form-message notice">
        <span v-if="isInElection">
          There is currently an ongoing election for new validator candidates.
          Stake is not allowed by now.
        </span>
        <span
          v-if="hasNoNominations && Object.keys(targetValidator).length === 0"
        >
          Before staking any tokens you need to select some validators to start
          getting rewards. Select one by going to its profile page and clicking
          on "Select"
        </span>
        <span v-else-if="!isRedelegation && !isNomination">
          It will take {{ undelegationPeriod }} to unlock your tokens after they
          are staked. There is a risk that some tokens will be lost depending on
          the behaviour of the validator you choose.
        </span>
        <span v-else-if="!isNomination && !isUnnomination">
          Voting power and rewards will change instantly upon restaking â€” but
          your tokens will still be subject to the risks associated with the
          original stake for the duration of the unstaking period.
        </span>
        <span v-else-if="isNomination">
          The validator will appear within your validator set once the next era
          begins.
        </span>
      </div>
    </TmFormGroup>
    <TmFormGroup
      v-if="
        Object.keys(targetValidator).length > 0 &&
        session.addressRole !== `stash`
      "
      class="action-modal-form-group"
      field-id="to"
      field-label="To"
    >
      <TmField id="to" :value="enhancedTargetValidator" type="text" readonly />
      <template>
        <TmFormMsg
          v-if="targetValidator.status === 'INACTIVE' && !isRedelegation"
          :msg="`You are about to stake to an inactive validator (${targetValidator.statusDetailed})`"
          type="custom"
          class="tm-form-msg--desc"
        />
        <TmFormMsg
          v-if="targetValidator.status === 'INACTIVE' && isRedelegation"
          :msg="`You are about to restake to an inactive validator (${targetValidator.statusDetailed})`"
          type="custom"
          class="tm-form-msg--desc"
        />
      </template>
    </TmFormGroup>

    <TmFormGroup
      v-if="
        Object.keys(targetValidator).length > 0 &&
        session.addressRole !== `stash`
      "
      class="action-modal-form-group"
      field-id="from"
      field-label="From"
    >
      <TmField
        id="from"
        v-model="fromSelectedIndex"
        :title="from"
        :options="fromOptions"
        type="select"
      />
    </TmFormGroup>
    <TmFormGroup
      v-if="
        currentNetwork.network_type === `polkadot`
          ? (!isNomination || stakedBalance.total === 0) &&
            session.addressRole !== `controller`
          : true
      "
      :error="$v.amount.$error && $v.amount.$invalid"
      class="action-modal-form-group"
      field-id="amount"
      :field-label="`Amount${
        currentNetwork.network_type === 'polkadot' &&
        Object.keys(targetValidator).length > 0 &&
        balance.total > 0 &&
        session.addressRole !== `stash`
          ? ' (Optional)'
          : ''
      }`"
    >
      <span class="input-suffix max-button">{{
        currentNetwork.stakingDenom
      }}</span>
      <TmFieldGroup>
        <TmField
          id="amount"
          v-model="amount"
          v-focus
          placeholder="0"
          class="tm-field-addon"
          type="number"
          @keyup.enter.native="enterPressed"
        />
        <TmBtn
          type="button"
          class="secondary addon-max"
          value="Set Max"
          :disabled="session.addressRole === `controller`"
          @click.native="setMaxAmount()"
        />
      </TmFieldGroup>
      <span class="form-message">
        Available to stake:
        {{ maxAmount }}
        {{ currentNetwork.stakingDenom }}s
      </span>
      <TmFormMsg
        v-if="balance.available === '0'"
        :msg="`doesn't have any ${currentNetwork.stakingDenom}s`"
        name="Wallet"
        type="custom"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.decimal"
        name="Amount"
        type="numeric"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && (!$v.amount.required || amount === 0)"
        name="Amount"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.max"
        type="custom"
        :msg="`You don't have enough ${currentNetwork.stakingDenom}s to proceed.`"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.min"
        :min="smallestAmount"
        name="Amount"
        type="min"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.maxDecimals"
        name="Amount"
        type="maxDecimals"
      />
      <TmFormMsg
        v-else-if="isMaxAmount() && !isRedelegation"
        msg="You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
        type="custom"
        class="tm-form-msg--desc"
      />
    </TmFormGroup>
  </ActionModal>
</template>

<script>
import { mapState, mapGetters } from "vuex"
import { decimal } from "vuelidate/lib/validators"
import gql from "graphql-tag"
import { SMALLEST } from "src/scripts/num"
import TmField from "src/components/common/TmField"
import TmFieldGroup from "src/components/common/TmFieldGroup"
import TmBtn from "src/components/common/TmBtn"
import TmFormGroup from "src/components/common/TmFormGroup"
import TmFormMsg from "src/components/common/TmFormMsg"
import ActionModal from "./ActionModal"
import { messageType } from "../../components/transactions/messageTypes"
import { formatAddress, validatorEntry } from "src/filters"
import { ValidatorFragment, UserTransactionAdded } from "src/gql"

export default {
  name: `delegation-modal`,
  components: {
    TmField,
    TmFieldGroup,
    TmBtn,
    TmFormGroup,
    TmFormMsg,
    ActionModal,
  },
  filters: {
    validatorEntry,
  },
  props: {
    targetValidator: {
      type: Object,
      default: () => ({}),
    },
    isNomination: {
      type: Boolean,
      default: false,
    },
    isUnnomination: {
      type: Boolean,
      default: false,
    },
  },
  data: () => ({
    amount: 0,
    fromSelectedIndex: 0,
    balance: {
      amount: null,
      denom: ``,
    },
    validators: [],
    delegations: [],
    undelegations: [],
    undelegationsLoaded: false,
    messageType,
    smallestAmount: SMALLEST,
    isInElection: false, // Handle election period in Polkadot
  }),
  computed: {
    ...mapState([`session`]),
    ...mapGetters([`network`, `address`, `currentNetwork`]),
    stakedBalance() {
      // balances not loaded yet
      if (!this.balance) {
        return {
          total: 0,
          denom: this.currentNetwork.stakingDenom,
        }
      }
      let stakedAmount =
        Number(this.balance.total) - Number(this.balance.available)
      // substract the already unbonding balance in the case of Substrate networks.
      if (this.undelegationsLoaded && this.undelegations.length > 0) {
        stakedAmount = this.undelegations.reduce(
          (stakedAmount, { amount }) => stakedAmount - Number(amount),
          stakedAmount
        )
      }
      return {
        total: Number(stakedAmount),
        denom: this.currentNetwork.stakingDenom,
      }
    },
    toOptions() {
      return this.validators
        .filter(
          (validator) =>
            validator.operatorAddress === this.targetValidator.operatorAddress
        )
        .map((validator) => {
          return {
            address: validator.operatorAddress,
            key: `${validator.name} - ${formatAddress(
              validator.operatorAddress
            )}`,
            value: 0,
          }
        })
    },
    fromOptions() {
      let options = [
        // from wallet
        {
          address: this.address,
          maximum: Number(this.balance.available),
          key: `My Wallet - ${formatAddress(this.address)}`,
          value: 0,
        },
      ]
      options = options.concat(
        this.delegations
          // exclude target validator
          .filter(
            (delegation) =>
              delegation.validator.operatorAddress !==
              this.targetValidator.operatorAddress
          )
          .map((delegation, index) => {
            return {
              address: delegation.validator.operatorAddress,
              maximum: Number(delegation.amount),
              key: validatorEntry(delegation.validator),
              value: index + 1,
            }
          })
      )

      return options
    },
    from() {
      if (!this.session.signedIn) return ``

      return this.fromOptions[this.fromSelectedIndex].address
    },
    transactionData() {
      if (isNaN(this.amount)) return {}

      if (this.isRedelegation) {
        return {
          type: messageType.RESTAKE,
          from: [this.from],
          to:
            Object.keys(this.targetValidator).length > 0
              ? [this.targetValidator.operatorAddress]
              : "",
          amount: {
            amount: this.amount,
            denom: this.currentNetwork.stakingDenom,
          },
          addressRole: this.session.addressRole,
        }
      } else {
        return {
          type: messageType.STAKE,
          to:
            Object.keys(this.targetValidator).length > 0
              ? [this.targetValidator.operatorAddress]
              : "",
          amount: {
            amount: this.amount,
            denom: this.currentNetwork.stakingDenom,
          },
          addressRole: this.session.addressRole,
        }
      }
    },
    notifyMessage() {
      if (this.isRedelegation) {
        return {
          title: `Successfully restaked!`,
          body: `You have successfully restaked your ${this.currentNetwork.stakingDenom}s`,
        }
      } else {
        return {
          title: `Successfully staked!`,
          body: `You have successfully staked your ${this.currentNetwork.stakingDenom}s`,
        }
      }
    },
    maxAmount() {
      return this.fromOptions[this.fromSelectedIndex].maximum
    },
    isRedelegation() {
      return this.fromSelectedIndex !== 0 && this.fromSelectedIndex !== "0" // where are these 0 strings comming from?
    },
    undelegationPeriod() {
      return this.currentNetwork.lockUpPeriod
    },
    enhancedTargetValidator() {
      return validatorEntry(this.targetValidator)
    },
    hasNoNominations() {
      // TODO: only temporary. Need to create a query to know if the address has any nomination
      return (
        this.currentNetwork.network_type === `polkadot` &&
        this.session.addressRole === `none`
      )
    },
  },
  methods: {
    open() {
      this.$refs.actionModal.open()
      this.$apollo.queries.balance.refetch()
      this.$apollo.queries.delegations.refetch()
    },
    validateForm() {
      this.$v.$touch()
      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()
      this.fromSelectedIndex = 0
      this.amount = 0
    },
    setMaxAmount() {
      this.amount = this.maxAmount
    },
    isMaxAmount() {
      return parseFloat(this.amount) === parseFloat(this.maxAmount)
    },
    enterPressed() {
      this.$refs.actionModal.validateChangeStep()
    },
    onSuccess(event) {
      this.$emit(`success`, event)

      // update registered topics for emails as the validator set changed
      this.$store.dispatch("updateNotificationRegistrations")
      // update the role of the user as it might change after bonding the first time
      if (this.currentNetwork.network_type === "polkadot") {
        this.$store.dispatch("checkAddressRole", {
          address: this.address,
          networkId: this.currentNetwork.id,
        })
      }
    },
  },
  validations() {
    return {
      amount: {
        required: (amount) => {
          // In Polkadot we don't need to bond extra, the user may just want to nominate a new validator
          // stash accounts or new accounts that haven't bonded tokens yet, need to specify an amount to bond
          if (
            this.currentNetwork.network_type === "polkadot" &&
            ["controller", "stash/controller"].includes(
              this.session.addressRole
            )
          ) {
            return true
          }
          return !!amount && amount !== `0`
        },
        decimal,
        max: (x) => Number(x) <= this.maxAmount,
        min: (x) => {
          // see required
          if (
            this.currentNetwork.network_type === "polkadot" &&
            ["controller", "stash/controller"].includes(
              this.session.addressRole
            )
          ) {
            return true
          }
          return Number(x) >= SMALLEST
        },
        maxDecimals: (x) => {
          return x.toString().split(".").length > 1
            ? x.toString().split(".")[1].length <= 6
            : true
        },
      },
    }
  },
  apollo: {
    validators: {
      query: gql`
        query validators($networkId: String!) {
          validators(networkId: $networkId) {
            id
            name
            operatorAddress
          }
        }
      `,
      /* istanbul ignore next */
      variables() {
        return {
          networkId: this.network,
        }
      },
      /* istanbul ignore next */
      update(data) {
        return data.validators
      },
      /* istanbul ignore next */
      skip() {
        return (
          !this.address ||
          !this.$refs.actionModal ||
          !this.$refs.actionModal.show
        )
      },
    },
    delegations: {
      query: gql`
        query delegations($networkId: String!, $delegatorAddress: String!) {
          delegations(
            networkId: $networkId
            delegatorAddress: $delegatorAddress
          ) {
            id
            amount
            validator {
              id
              operatorAddress
              name
            }
          }
        }
      `,
      /* istanbul ignore next */
      skip() {
        return (
          !this.address ||
          !this.$refs.actionModal ||
          !this.$refs.actionModal.show
        )
      },
      /* istanbul ignore next */
      variables() {
        return {
          networkId: this.network,
          delegatorAddress: this.address,
        }
      },
      /* istanbul ignore next */
      update(data) {
        return data.delegations
      },
    },
    balance: {
      query: gql`
        query balance($networkId: String!, $address: String!) {
          balancesV2(networkId: $networkId, address: $address) {
            total
            available
            denom
          }
        }
      `,
      /* istanbul ignore next */
      skip() {
        return (
          !this.address ||
          !this.currentNetwork.stakingDenom ||
          !this.$refs.actionModal ||
          !this.$refs.actionModal.show
        )
      },
      /* istanbul ignore next */
      variables() {
        return {
          networkId: this.network,
          address: this.address,
        }
      },
      /* istanbul ignore next */
      update(data) {
        return (
          data.balancesV2.find(
            ({ denom }) => denom === this.currentNetwork.stakingDenom
          ) || {
            amount: 0,
          }
        )
      },
    },
    undelegations: {
      query() {
        /* istanbul ignore next */
        return gql`
        query undelegations($networkId: String!, $delegatorAddress: String!) {
          undelegations(networkId: $networkId, delegatorAddress: $delegatorAddress) {
            id
            validator {
              ${ValidatorFragment}
            }
            amount
            startHeight
            endTime
          }
        }
      `
      },
      /* istanbul ignore next */
      variables() {
        return {
          networkId: this.currentNetwork.id,
          delegatorAddress: this.address,
        }
      },
      /* istanbul ignore next */
      skip() {
        return this.currentNetwork.network_type !== `polkadot` // we only need undelegations for Polkadot networks
      },
      /* istanbul ignore next */
      update(data) {
        this.undelegationsLoaded = true
        return data.undelegations
      },
    },
    $subscribe: {
      userTransactionAdded: {
        /* istanbul ignore next */
        variables() {
          return {
            networkId: this.network,
            address: this.address,
          }
        },
        /* istanbul ignore next */
        skip() {
          return (
            !this.address ||
            !this.$refs.actionModal ||
            !this.$refs.actionModal.show
          )
        },
        query: UserTransactionAdded,
        /* istanbul ignore next */
        result() {
          this.$apollo.queries.balance.refetch()
          this.$apollo.queries.delegations.refetch()
        },
      },
      blockAdded: {
        /* istanbul ignore next */
        variables() {
          return {
            networkId: this.network,
          }
        },
        /* istanbul ignore next */
        query() {
          return gql`
            subscription blockAdded($networkId: String!) {
              blockAdded(networkId: $networkId) {
                data
              }
            }
          `
        },
        /* istanbul ignore next */
        skip() {
          return this.currentNetwork.network_type !== "polkadot"
        },
        /* istanbul ignore next */
        result({ data }) {
          if (data.blockAdded.data) {
            this.isInElection = JSON.parse(data.blockAdded.data).isInElection
          }
        },
      },
    },
  },
}
</script>
