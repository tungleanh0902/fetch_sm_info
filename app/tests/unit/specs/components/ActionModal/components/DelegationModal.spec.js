"use strict"

import { shallowMount, createLocalVue } from "@vue/test-utils"
import DelegationModal from "src/ActionModal/components/DelegationModal"
import Vuelidate from "vuelidate"

const validators = [
  {
    operatorAddress: "cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctqzh8yqw",
    status: "ACTIVE",
  },
  {
    operatorAddress: "cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au",
    status: "ACTIVE",
  },
  {
    operatorAddress: "cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctgurrg7n",
    status: "INACTIVE",
    statusDetailed: "temporally banned from the network",
  },
  {
    operatorAddress: "cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctgurrg7m",
    status: "INACTIVE",
    statusDetailed: "banned from the network",
  },
]

describe(`DelegationModal`, () => {
  let wrapper
  const localVue = createLocalVue()
  localVue.use(Vuelidate)
  localVue.directive("focus", () => {})

  const state = {
    session: {
      signedIn: true,
      addressRole: undefined,
    },
  }

  const getters = {
    network: "testnet",
    currentNetwork: {
      id: "testnet",
      coinLookup: [
        { viewDenom: "STAKE", chainToViewConversionFactor: 0.000001 },
      ],
      lockUpPeriod: `21 days`,
      stakingDenom: "STAKE",
    },
    address: "cosmos1234",
  }

  beforeEach(() => {
    wrapper = shallowMount(DelegationModal, {
      localVue,
      mocks: {
        $store: { getters, state },
        $apollo: {
          queries: {
            balance: { refetch: () => {} },
            delegations: { refetch: () => {} },
            validators: { refetch: () => {} },
          },
        },
      },
      propsData: {
        targetValidator: validators[0],
      },
    })
    wrapper.setData({
      delegations: [
        {
          validator: validators[0],
          amount: 10,
        },
        {
          validator: validators[1],
          amount: 124,
        },
        {
          validator: validators[2],
          amount: 200,
        },
      ],
      balance: {
        total: 2000,
        available: 1000,
        denom: "STAKE",
      },
      validators: validators,
    })
  })

  it(`should display the delegation modal form`, async () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should submit when enterPressed is called`, async () => {
    const self = {
      $refs: { actionModal: { validateChangeStep: jest.fn() } },
    }
    DelegationModal.methods.enterPressed.call(self)
    expect(self.$refs.actionModal.validateChangeStep).toHaveBeenCalled()
  })

  it(`opens`, () => {
    const $refs = { actionModal: { open: jest.fn() } }
    const $apollo = wrapper.vm.$apollo
    DelegationModal.methods.open.call({ $refs, $apollo })
    expect($refs.actionModal.open).toHaveBeenCalled()
  })

  it(`clears on close`, () => {
    const self = {
      $v: { $reset: jest.fn() },
      fromSelectedIndex: 1,
      amount: 10,
    }
    DelegationModal.methods.clear.call(self)
    expect(self.$v.$reset).toHaveBeenCalled()
    expect(self.fromSelectedIndex).toBe(0)
    expect(self.amount).toBe(0)
  })

  describe(`if amount field max button clicked`, () => {
    it(`amount has to be 1000 atom`, async () => {
      wrapper.vm.setMaxAmount()
      expect(wrapper.vm.amount).toBe(1000)
    })
  })

  describe(`validation`, () => {
    describe(`fails`, () => {
      it(`with default values`, () => {
        expect(wrapper.vm.validateForm()).toBe(false)
      })

      it(`if the user manually inputs a number greater than the balance`, () => {
        wrapper.setData({ amount: 1420 })
        expect(wrapper.vm.validateForm()).toBe(false)
      })
    })

    describe(`succeeds`, () => {
      it(`if the amount is positive and the user has enough balance`, () => {
        wrapper.setData({ amount: 50 })
        expect(wrapper.vm.validateForm()).toBe(true)
      })
    })
  })

  describe("Submission Data for Delegating", () => {
    beforeEach(() => {
      wrapper.setProps({
        targetValidator: validators[1],
      })
      wrapper.setData({
        amount: 10,
        fromSelectedIndex: 0,
      })
    })

    it("should return correct transaction data for delegating", () => {
      expect(wrapper.vm.transactionData).toEqual({
        type: "StakeTx",
        amount: {
          amount: 10,
          denom: "STAKE",
        },
        to: ["cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au"],
      })
    })

    it("should return empty transaction data if amount is NaN", () => {
      wrapper.setData({
        amount: `NaN`,
      })
      expect(wrapper.vm.transactionData).toEqual({})
    })

    it("should return correct notification message for delegating", () => {
      expect(wrapper.vm.notifyMessage).toEqual({
        title: `Successfully staked!`,
        body: `You have successfully staked your STAKEs`,
      })
    })
  })

  describe("Submission Data for Redelegating", () => {
    beforeEach(() => {
      wrapper.setProps({
        targetValidator: validators[2],
      })
      wrapper.setData({
        amount: 10,
        fromSelectedIndex: 2,
      })
    })

    it("should set the subtotal to 0 on a restake", () => {
      expect(wrapper.html()).toMatchSnapshot()
    })

    it("should return correct transaction data for redelegating", () => {
      expect(wrapper.vm.transactionData).toEqual({
        type: "RestakeTx",
        amount: {
          amount: 10,
          denom: "STAKE",
        },
        from: ["cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au"],
        to: ["cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctgurrg7n"],
      })
    })

    it("should return correct notification message for delegating", () => {
      expect(wrapper.vm.notifyMessage).toEqual({
        title: `Successfully restaked!`,
        body: `You have successfully restaked your STAKEs`,
      })
    })

    it(`should send an event on success`, () => {
      const self = {
        $emit: jest.fn(),
        $store: {
          dispatch: jest.fn(),
        },
        currentNetwork: {
          network_type: "polkadot",
        },
      }
      DelegationModal.methods.onSuccess.call(self)
      expect(self.$emit).toHaveBeenCalledWith(
        "success",
        expect.objectContaining({})
      )
    })

    it(`re register notifications on success`, () => {
      const self = {
        $emit: jest.fn(),
        $store: {
          dispatch: jest.fn(),
        },
        currentNetwork: {
          network_type: "polkadot",
        },
      }
      DelegationModal.methods.onSuccess.call(self)
      expect(self.$store.dispatch).toHaveBeenCalledWith(
        "updateNotificationRegistrations"
      )
    })
  })

  describe(`if amount field max button clicked`, () => {
    it(`amount has to be 1000 atom`, async () => {
      wrapper.setData({
        amount: 1,
        fromSelectedIndex: `0`,
      })
      wrapper.vm.setMaxAmount()
      expect(wrapper.vm.amount).toBe(1000)
    })
    it(`should show warning message`, async () => {
      wrapper.setData({
        amount: 1000,
        fromSelectedIndex: `0`,
      })
      //await wrapper.vm.$nextTick()
      expect(wrapper.html()).toContain(
        "You are about to use all your tokens for this transaction. Consider leaving a little bit left over to cover the network fees."
      )
    })
  })

  describe(`if validator is jailed`, () => {
    it(`must show warn message about it`, async () => {
      wrapper.setProps({
        targetValidator: validators[2], // Jailed validator
      })
      wrapper.setData({
        amount: 1,
        fromSelectedIndex: `0`,
      })
      expect(wrapper.html()).toContain(
        "You are about to stake to an inactive validator (temporally banned from the network)"
      )
    })
  })

  describe(`if validator is tombstoned`, () => {
    it(`must show warn message about it`, async () => {
      wrapper.setProps({
        targetValidator: validators[3], // Tombstoned validator
      })
      wrapper.setData({
        amount: 1,
        fromSelectedIndex: `0`,
      })
      expect(wrapper.html()).toContain(
        "You are about to stake to an inactive validator (banned from the network)"
      )
    })
  })

  describe(`if validator is active`, () => {
    it(`must not show warn message`, async () => {
      wrapper.setData({
        amount: 1,
        fromSelectedIndex: `0`,
        targetValidator: validators[0], // Active validator
      })
      expect(wrapper.html()).not.toContain(
        "You are about to stake to an inactive validator"
      )
      expect(wrapper.find("#to .tm-form-msg--desc").exists()).toBe(false)
    })
  })

  describe(`if network is polkadot`, () => {
    it(`amount should be required if address is stash`, async () => {
      wrapper.vm.$store.state.session.addressRole = `stash`
      wrapper.vm.$store.getters.currentNetwork.network_type = "polkadot"
      wrapper.setData({
        amount: 0,
        fromSelectedIndex: `0`,
        targetValidator: validators[0], // Active validator
      })
      expect(wrapper.vm.validateForm()).toBe(false)
    })
    it(`amount should not be required if address is controller`, async () => {
      wrapper.vm.$store.state.session.addressRole = `controller`
      wrapper.vm.$store.getters.currentNetwork.network_type = "polkadot"
      wrapper.setData({
        amount: 0,
        fromSelectedIndex: `0`,
        targetValidator: validators[0], // Active validator
      })
      expect(wrapper.vm.validateForm()).toBe(true)
    })
  })
})
