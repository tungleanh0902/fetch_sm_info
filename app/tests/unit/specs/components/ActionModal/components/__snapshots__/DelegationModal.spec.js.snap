// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DelegationModal Submission Data for Redelegating should set the subtotal to 0 on a restake 1`] = `
<actionmodal-stub title="Restake" validate="function () { [native code] }" submissionerrorprefix="Staking failed" amount="0" rewards="" transactiondata="[object Object]" notifymessage="[object Object]" featureflag="delegate" selecteddenom="" transactiontype="RestakeTx" id="delegation-modal" class="delegation-modal">
  <!---->
  <!---->
  <tmformgroup-stub class="action-modal-form-group">
    <div class="form-message notice">
      <!----> <span>
        Voting power and rewards will change instantly upon restaking — but
        your tokens will still be subject to the risks associated with the
        original stake for the duration of the unstaking period.
      </span></div>
  </tmformgroup-stub>
  <tmformgroup-stub fieldid="to" fieldlabel="To" class="action-modal-form-group">
    <tmfield-stub type="text" value="undefined - cosmosvaladdr…3p4mqpjyrm5ctgurrg7n" id="to" readonly=""></tmfield-stub>
    <!---->
    <tmformmsg-stub type="custom" name="" msg="You are about to restake to an inactive validator (temporally banned from the network)" class="tm-form-msg--desc"></tmformmsg-stub>
  </tmformgroup-stub>
  <tmformgroup-stub fieldid="from" fieldlabel="From" class="action-modal-form-group">
    <tmfield-stub type="select" value="2" options="[object Object],[object Object],[object Object]" id="from" title="cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctplpn3au"></tmfield-stub>
  </tmformgroup-stub>
  <tmformgroup-stub fieldid="amount" fieldlabel="Amount" class="action-modal-form-group"><span class="input-suffix max-button">STAKE</span>
    <tmfieldgroup-stub>
      <tmfield-stub type="number" value="10" placeholder="0" id="amount" class="tm-field-addon"></tmfield-stub>
      <tmbtn-stub value="Set Max" type="button" customclass="" class="secondary addon-max"></tmbtn-stub>
    </tmfieldgroup-stub> <span class="form-message">
      Available to stake:
      124
      STAKEs
    </span>
    <!---->
  </tmformgroup-stub>
</actionmodal-stub>
`;

exports[`DelegationModal should display the delegation modal form 1`] = `
<actionmodal-stub
  amount="0"
  class="delegation-modal"
  featureflag="delegate"
  id="delegation-modal"
  notifymessage="[object Object]"
  rewards=""
  selecteddenom=""
  submissionerrorprefix="Staking failed"
  title="Stake"
  transactiondata="[object Object]"
  transactiontype="StakeTx"
  validate="function () { [native code] }"
>
  <!---->
   
  <!---->
   
  <tmformgroup-stub
    class="action-modal-form-group"
  >
    <div
      class="form-message notice"
    >
      <!---->
       
      <span>
        
        It will take 21 days to unlock your tokens after they
        are staked. There is a risk that some tokens will be lost depending on
        the behaviour of the validator you choose.
      
      </span>
    </div>
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="to"
    fieldlabel="To"
  >
    <tmfield-stub
      id="to"
      readonly=""
      type="text"
      value="undefined - cosmosvaladdr…3p4mqpjyrm5ctqzh8yqw"
    />
     
    <!---->
     
    <!---->
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="from"
    fieldlabel="From"
  >
    <tmfield-stub
      id="from"
      options="[object Object],[object Object],[object Object]"
      title="cosmos1234"
      type="select"
      value="0"
    />
  </tmformgroup-stub>
   
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
  >
    <span
      class="input-suffix max-button"
    >
      STAKE
    </span>
     
    <tmfieldgroup-stub>
      <tmfield-stub
        class="tm-field-addon"
        id="amount"
        placeholder="0"
        type="number"
        value="0"
      />
       
      <tmbtn-stub
        class="secondary addon-max"
        customclass=""
        type="button"
        value="Set Max"
      />
    </tmfieldgroup-stub>
     
    <span
      class="form-message"
    >
      
      Available to stake:
      1000
      STAKEs
    
    </span>
     
    <!---->
  </tmformgroup-stub>
</actionmodal-stub>
`;
